### ISSUE TITLE
Global automount configuration

### ISSUE THREAD

**Issue by vladimir-novoseltsev** [microsoft/WSL#7236]
Hello, please add an option to globally configure automount with .wslconfig file.
Option to not automount removable drives would be perfect as well, right now, if someone use removable drives they cannot be safely removed until WSL distros with automount enabled are shutdown.

**Comment by WSLUser** [comment]
is `/etc/fstab` not enough of a global automount for you?

**Comment by vladimir-novoseltsev** [comment]
`/etc/fstab` is per distro, so it is not global and it is not amount at all.

**Comment by vladimir-novoseltsev** [comment]
Ok, sorry, I've probably been not precise enough in what I wanted.
Right now you can set per distro if host's drives are mounted when WSL distro is started. By default all host's drives are mounted into distro.
You can disable that with `/etc/wsl.conf` with following:
```
[automount]
enabled=false
```
Then you can mount whatever you want through `/etc/fstab`.

The problem I see is this behavior mount ALL drives from the host system, including removable drives, which you cannot safely remove without stopping WSL distros. This is ok, as long as you manage and configure it yourself. Sometime you don't want to mess with distro config, like i.e. Docker WSL2 distro as it will loose some flexibility and might be broken in result.

So what I want is option to globally for all WSL distros to disable mounting for host's removable drives.

**Comment by WSLUser** [comment]
Interesting. Many people wanted precisely the behavior of auto-mounting removable drives. To the point projects exist using things like USB-IP for WSL2 to pick up on them.  You didn't specify, but sounds like based on what I just said, you're using WSL1 and want more options available under [automount] to be configurable. I think the answer from the MS devs will still be use `/etc/fstab`, which only mounts what's defined. You can copy that to every distro, they're all the same when it comes to mounting. I suppose they could make a global /etc/fstab that's written to every distro you install but they prefer to remain hands off of the distros as much as possible. It's bad enough we still have a proprietary `init `being used on WSL2.

**Comment by dinvlad** [comment]
I have another use case for this - to isolate WSL distros from Windows for more sensitive scenarios. For example, I want to prevent malware downloaded inside WSL from accessing (or overriding access to) the Windows host. Right now, I can disable it via `wsl.conf`, but that _can_ be overridden by malware (assuming it got root inside WSL). It would be nice to be able to enforce `[automount] enabled=false` globally instead.

**Comment by cycledetector** [comment]
Has there been any progress towards implementing this feature? 
Being able to set [automount] enabled = false and [interop] enabled = false in .wslconfig, i.e. globally, instead of setting those options in wsl.conf per distro would actually be a very useful feature.
As dinvlad said, if we assume that someone has root access inside a WSL distro, wsl.conf can be modified, while modifying .wslconfig would require access to the actual Windows host.

**Comment by Biswa96** [comment]
It may be possible with a little Win32 program. The logic would be like following:
1. Use [`WslGetDistributionConfiguration()`][1] to get wslDistributionFlags.
2. Remove the [flags][2] you want to disable.
3. Set the remaining flag value with [`WslConfigureDistribution()`][3].

[1]: https://docs.microsoft.com/en-us/windows/win32/api/wslapi/nf-wslapi-wslgetdistributionconfiguration
[2]: https://docs.microsoft.com/en-us/windows/win32/api/wslapi/ne-wslapi-wsl_distribution_flags
[3]: https://docs.microsoft.com/en-us/windows/win32/api/wslapi/nf-wslapi-wslconfiguredistribution

**Comment by cycledetector** [comment]
> It may be possible with a little Win32 program. The logic would be like following:
> 
> 1. Use [`WslGetDistributionConfiguration()`](https://docs.microsoft.com/en-us/windows/win32/api/wslapi/nf-wslapi-wslgetdistributionconfiguration) to get wslDistributionFlags.
> 2. Remove the [flags](https://docs.microsoft.com/en-us/windows/win32/api/wslapi/ne-wslapi-wsl_distribution_flags) you want to disable.
> 3. Set the remaining flag value with [`WslConfigureDistribution()`](https://docs.microsoft.com/en-us/windows/win32/api/wslapi/nf-wslapi-wslconfiguredistribution).

As far as I understand, this flag value is set in the Windows Registry. However, someone can still mount a host drive and enable interoperability by creating the /etc/wsl.conf file, right?

**Comment by cycledetector** [comment]
> > However, someone can still mount a host drive and enable interoperability by modifying the /etc/wsl.conf file or manually mounting a drive, right?
> 
> As for the formality, I do not suggest to modify registry value because any wrong step may break your WSL setup. But _be as naughty as you want, just don’t get caught_. Yes, those flags are set in `Flags` registry. Those will override any options in /etc/wsl.conf. For example, if `WSL_DISTRIBUTION_FLAGS_ENABLE_DRIVE_MOUNTING` flag is removed `[automount] enabled=true` would do nothing. After that, Windows drives will **not** be automatically mounted inside of WSL sessions.

Thanks for the clarification, that's a good workaround for stopping automounting and blocking interopability I guess, but unless I am missing something, someone could still mount the host drives manually, is that correct?

**Comment by Biswa96** [comment]
> someone could still mount the host drives manually, is that correct?

Only if the current user has administrator privilege or belongs to Hyper-V Administrators user group.

**Comment by cycledetector** [comment]
> > someone could still mount the host drives manually, is that correct?
> 
> Only if the current user has administrator privilege or belongs to Hyper-V Administrators user group.

I found out that it is possible to also mount a physical drive from within a Linux distro running on WSL. I disabled automount and interoperability and executed: `sudo mount -t drvfs e: /mnt/e`.
The drive was properly mounted at /mnt/e and I could list its contents.
So while `wsl --mount` has to be run from the host (assuming that interoperability and automounting is off), if we assume that someone has root access on the Linux distro, access to the physical drives of the host is still possible.

I believe it would be useful to be able to restrict access to certain drives, especially now that WSL 2 is available on Windows Server. I understand that WSL is a product geared towards development workstations and not production servers but on the other hand, not being to limit access to certain resources is a security oversight in my humble opinion.
In the same vain, I think that wsl.conf should not be located in the distro. As you mentioned, configuration for each Linux distro already exists in the Windows Registry, so what is the purpose of having a second one, one that can be modified from inside the distro?

**Comment by cycledetector** [comment]
> > I disabled automount and interoperability and executed: sudo mount -t drvfs e: /mnt/e.
> 
> Did you disable that with those WSL APIs? I have tried that command and it shows error like the following after disabling `WSL_DISTRIBUTION_FLAGS_ENABLE_DRIVE_MOUNTING` flag.
> 
> ```
> $ sudo mount -t drvfs e: /mnt/e
> <3>init: (23) ERROR: UtilConnectVsock:408: connect port 50002 failed 110
> Operation timed out
> ```

I haven't tried using the WSL API. I disabled mounting through /etc/wsl.conf initially.
I tried again by setting the Flags registry value to `8` in `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Lxss\` for the distro. I assume this to be equivalent to using the Wslapi.
I set it to `8` since this is flag for defining WSL version 2 without any additional permissions. I assume `9` would be the value to only enable interoperability, `a` to only append Windows path elements to $PATH, `b`  to do both,  `c` to only enable drive mounting and `f` to enable everything.

When you changed WSL_DISTRIBUTION_FLAGS_ENABLE_DRIVE_MOUNTING, did the version of the WSL distro remain 2? Because anything under `8` for the flag value sets the WSL distro to version 1.

**Comment by Biswa96** [comment]
I have tried that with both WSL1 and WSL2. Here I focused on the automount feature. Whether `Flags` works with different values is a broad topic. Some calculation,
* WSL_DISTRIBUTION_FLAGS_ENABLE_DRIVE_MOUNTING = 0x4
* WSL1: Before 0x7 → Subtract 0x4 → After 0x3
* WSL2: Before 0xf → Subtract 0x4 → After 0xb

**Comment by thawkins** [comment]
This is an important requirement for us to be able to permenantly and globaly disable interop and automount and also disable drvfs completly, without non local admins or wsl sudo users being able to enable it again.. We are a large corporate, and our security division will not allow the deployment of wsl without this ability to lock it down.

**Comment by dinvlad** [comment]
This is pretty much the reason why I stopped using WSL, unfortunately. It worked pretty flawlessly otherwise, but the lack of separation between the host and guests is just not workable in a multi-project environment.

**Comment by Detzler** [comment]
> This is pretty much the reason why I stopped using WSL, unfortunately. It worked quite flawlessly otherwise, but the lack of separation between the host and guests is just not workable in a multi-project environment.

I fully agree. Had the same situation and consequence. Can't use WSL in my enterprise field because of this security issue.
